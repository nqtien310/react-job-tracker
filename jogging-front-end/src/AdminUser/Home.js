import React from 'react';
import UsersList from '../Users'
import './home.css'

export default function Home(props) {
  return <UsersList />
}
