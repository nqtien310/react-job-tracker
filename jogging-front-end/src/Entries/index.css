table td {
  width: 180px;
}

table td input.form-control{
  width: 100px;
}

.field-container {
  width: 130px;
  margin-right: 50px;
  display: inline-block;
}

.date-filters {
  margin-bottom: 20px
}

.date-filters > div {
  display: inline-block;
}
.date-filter {
  width: 150px;
  margin-left: 30px;
}
