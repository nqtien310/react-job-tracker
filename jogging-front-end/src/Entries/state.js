import Rx from 'rxjs/Rx'
import api from '../api'
import { combineReducers } from 'redux'
import { push } from 'react-router-redux'
import setErrorMessage from '../actions/setErrorMessage.js'
import {reset} from 'redux-form'
//constants
export const ENTRY_FORM_CREATE = 'ENTRY_FORM_CREATE'
export const ENTRY_FORM_EDIT   = 'ENTRY_FORM_EDIT'
export const ENTRY_FORM_HIDE   = 'ENTRY_FORM_HIDE'
export const ENTRY_DELETE      = 'ENTRY_DELETE'
export const ENTRY_CREATE      = 'ENTRY_CREATE'
export const ENTRY_UPDATE      = 'ENTRY_UPDATE'
export const ENTRIES_FETCH     = 'ENTRIES_FETCH'
export const ENTRIES_FETCHED   = 'ENTRIES_FETCHED'

//actions
export const deleteEntry = (userId, entryId) => {
  return {
    type: ENTRY_DELETE,
    payload:{
      userId: userId,
      id: entryId
    }
  }
}

export const fetchEntries = (userId) => ({type: ENTRIES_FETCH, payload: userId})
export const showCreateForm = () => ({type: ENTRY_FORM_CREATE})
export const showEditForm = (entry) => ({type: ENTRY_FORM_EDIT, payload: entry})
export const hideForm = (entry) => ({type: ENTRY_FORM_HIDE})
export const createEntry = (userId, params) => {
  return {
    type: ENTRY_CREATE,
    payload: {
      userId: userId,
      params: params
    }
  }
}

export const updateEntry = (userId, id, params) => {
  return {
    type: ENTRY_UPDATE,
    payload: {
      userId: userId,
      id: id,
      params: params
    }
  }
}

export const fetchedEntries = (response) => {
  return {
    type: ENTRIES_FETCHED,
    payload: response.data
  }
}

//reducers
function listReducer(state=[], action){
  switch(action.type){
    case ENTRIES_FETCHED: {
      return action.payload
    }default: {
      return state
    }
  }
}

function formReducer(state={}, action){
  switch(action.type){
    case ENTRY_FORM_CREATE: {
      return {action: 'create'}
    }case ENTRY_FORM_EDIT: {
      return {action: 'edit', editing: action.payload}
    }case ENTRY_FORM_HIDE: {
      return {}
    }default: {
      return state
    }
  }
}

export const entryReducer = combineReducers({
  list: listReducer,
  form: formReducer
})

//Epics

export const deleteEntryEpic = (action$) => {
  return action$.ofType(ENTRY_DELETE).switchMap( action => {
    return Rx.Observable.fromPromise(
      api.delete(`users/${action.payload.userId}/entries/${action.payload.id}`)
    ).flatMap(response => [fetchEntries(action.payload.userId), hideForm()])
  })
}

export const updateEntryEpic = (action$) => {
  return action$.ofType(ENTRY_UPDATE).switchMap( action => {
    return Rx.Observable.fromPromise(
      api.put(`users/${action.payload.userId}/entries/${action.payload.id}`, action.payload.params)
    ).flatMap(response => [fetchEntries(action.payload.userId), hideForm()])
  })
}

export const fetchEntriesEpic = (action$) => {
  return action$.ofType(ENTRIES_FETCH).switchMap( action => {
    return Rx.Observable.fromPromise(
      api.get(`users/${action.payload}/entries`)
    ).map(response => fetchedEntries(response))
  })
}

export const createEntryEpic = (action$) => {
  return action$.ofType(ENTRY_CREATE).switchMap( action => {
    return Rx.Observable.fromPromise(
      api.post(`users/${action.payload.userId}/entries`, action.payload.params)
    ).flatMap(() => [fetchEntries(action.payload.userId), hideForm()]
    ).catch(error => Rx.Observable.of(setErrorMessage(error)))
  })
}


//https://github.com/redux-observable/redux-observable/tree/master/examples/navigation/epics
