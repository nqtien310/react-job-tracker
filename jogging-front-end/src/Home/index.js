import React from 'react';
import Template from '../Template'
import { connect } from 'react-redux'
import NonLoginHome from './NonLoginHome'
import RegularUserHome from '../RegularUser/Home'
import AdminUserHome from '../AdminUser/Home'

function Home(props) {
  function renderHomeAccordingly(){
    if(!props.myUser) {
      return <NonLoginHome/>
    }else if(props.myUser.role === 'user') {
      return <RegularUserHome myUser={props.myUser}/>
    }else if(props.myUser.role === 'admin') {
      return <AdminUserHome myUser={props.myUser}/>
    }
  }

  return (
    <Template> {renderHomeAccordingly()} </Template>
  )
}

function mapStateToProps(state){
  return {
    myUser: state.myUser
  }
}

export default connect(mapStateToProps)(Home)
