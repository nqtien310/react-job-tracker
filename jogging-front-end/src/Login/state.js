import { fetchMyUser } from '../epics/fetchMyUserEpic'
import Auth from '../auth'
import 'rxjs/add/operator/switchMap';
import Rx from 'rxjs/Rx'
import api from '../api'
import setErrorMessage from '../actions/setErrorMessage.js'

export const LOGIN  = 'LOGIN'
export const SETTOKEN = 'SETTOKEN'

export const submitLogin = params => ({type: LOGIN, payload: params})
export const setToken    = params => ({type: SETTOKEN, payload: params})


export const loginEpic = (action$) => {
  return action$.ofType(LOGIN).switchMap(action =>{
    return Rx.Observable.fromPromise(
      api.post('login', {user: action.payload})
    ).flatMap(response =>
        [setToken(response), fetchMyUser()]
      ).catch(error => setErrorMessage(error))
  })
}

