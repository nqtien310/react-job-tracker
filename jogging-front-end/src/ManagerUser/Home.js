import React from 'react';
import UsersList from '../Users'

export default function Home(props) {
  return <UsersList myUser={props.myUser}/>
}
