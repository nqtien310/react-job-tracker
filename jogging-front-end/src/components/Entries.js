import React from 'react';
import { connect } from 'react-redux'
import { deleteEntry } from '../epics/deleteEntryEpic'

class Entries extends React.Component{
  constructor(props) {
    super(props)
  }

  onDelete(id){
    this.props.deleteEntry(this.props.userId, id)
  }

  renderEntries() {
    return this.props.entries.map(entry => {
      return(
        <tr key={entry.id}>
          <td>
            {entry.distance_in_metre} m
          </td>
          <td>
            {entry.speed} m/s
          </td>
          <td>
            {entry.time_in_second} s
          </td>
          <td>
            {entry.formatted_date}
          </td>
          <td>
            <a onClick={()=> this.onDelete(entry.id)} className="btn btn-danger">Delete</a>
          </td>
        </tr>
      )
    })
  }

  render(){
    return (
      <table className="table">
        <thead>
          <tr>
            <th>Distance</th>
            <th>Speed</th>
            <th>Time In Second</th>
            <th>Date</th>
            <th>Action</th>
          </tr>
        </thead>

        <tbody>
          {this.renderEntries()}
        </tbody>
      </table>
    )
  }
}

function mapStateToProps(state){
  return {
    entries: state.entries
  }
}
export default connect(mapStateToProps, {
  deleteEntry
})(Entries)
