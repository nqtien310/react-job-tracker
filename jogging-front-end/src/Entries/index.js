import React from 'react';
import { connect } from 'react-redux'
import { deleteEntry, showCreateForm } from './state'
import List from './list'
import Form from './form'
import ErrorMessage from '../components/ErrorMessage'
import './index.css'
import DatePicker from 'react-datepicker'
import { Field } from 'redux-form'
import moment from 'moment'
import { Link } from 'react-router-dom'
import 'react-datepicker/dist/react-datepicker.css'

class Entries extends React.Component{
  constructor(props) {
    super(props)
    this.state = {
      from: null,
      to: null
    }
  }

  handleChange(type, date) {
    this.setState({[type]: date})
  }

  filterFromTo(){
    let entries = this.props.entry.list
    let format  = "YYYY-MM-DD"

    if(this.state.from){
      entries = entries.filter(e =>{
        return moment(e.date, format) >= this.state.from
      })
    }

    if(this.state.to){
      entries = entries.filter(e =>{
        return moment(e.date, format) <= this.state.to
      })
    }

    return entries
  }

  render(){
    let entries = this.filterFromTo()

    return (
      <div>
        <div className="date-filters">
          Filters:
          <DatePicker placeholderText="From.."
            className="form-control date-filter"
            onChange={date => this.handleChange("from", date)}
            selected={this.state.from} />

          <DatePicker placeholderText="To.."
            reportLink={this.props.reportLink}
            className="form-control date-filter"
            onChange={date => this.handleChange("to", date)}
            selected={this.state.to} />
        </div>

        <ErrorMessage />
        <List userId={this.props.userId} entries={entries}/>
        {this.props.entry.form.action ? <Form initialValues={this.props.entry.form.editing} userId={this.props.userId}/> : null}

        <Link to={this.props.reportLink}>Report by week</Link>
        <div className="action-footers">
          <a onClick={this.props.showCreateForm} className="btn btn-primary">
            <i className="fa fa-plus"/>
          </a>
        </div>
      </div>
    )
  }
}

function mapStateToProps(state){
  return {
    entry: state.entry
  }
}
export default connect(mapStateToProps, {
  deleteEntry,
  showCreateForm
})(Entries)
