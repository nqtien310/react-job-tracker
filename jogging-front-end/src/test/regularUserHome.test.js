import React from 'react';
import renderer from 'react-test-renderer'
import LoginForm from '../Login'
import RenderComponent from './RenderComponent'
import { store } from './RenderComponent'
import Home from '../RegularUser/Home'
import { fetchEntries } from '../Entries/state'
import { stubApiSuccess } from './ApiStubs'

it('Regular User Home (No Entries)', (done) => {
  stubApiSuccess("get", [])

  const component = renderer.create(
    <RenderComponent>
      <Home myUser={{id: 1}}/>
    </RenderComponent>
  )

  setTimeout(() => {
    expect(component).toMatchSnapshot()
    done()
  },50)
});

it('Regular User Home (With Entries)', (done) => {
  let data = [
    {"id":42,"time_in_second":100,"distance_in_metre":100.0,"formatted_date":"2017-09-03","speed":1.0,"date":"2017-09-03"},
    {"id":43,"time_in_second":200,"distance_in_metre":200.0,"formatted_date":"2017-09-06","speed":1.0,"date":"2017-09-06"}
  ]

  stubApiSuccess("get", data)
  const component = renderer.create(
    <RenderComponent>
      <Home myUser={{id: 1}}/>
    </RenderComponent>
  )

  setTimeout(() => {
    expect(component).toMatchSnapshot()
    done()
  },50)
});
