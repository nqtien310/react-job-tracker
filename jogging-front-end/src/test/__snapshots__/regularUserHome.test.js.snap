// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Regular User Home (No Entries) 1`] = `
<div>
  <div>
    <div
      className="date-filters"
    >
      Filters:
      <div>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="form-control date-filter"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="From.."
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="form-control date-filter"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="To.."
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div />
    <div>
      No Entries created
    </div>
    <a
      href="/report"
      onClick={[Function]}
    >
      Report by week
    </a>
    <div
      className="action-footers"
    >
      <a
        className="btn btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-plus"
        />
      </a>
    </div>
  </div>
</div>
`;

exports[`Regular User Home (With Entries) 1`] = `
<div>
  <div>
    <div
      className="date-filters"
    >
      Filters:
      <div>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="form-control date-filter"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="From.."
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div>
        <div
          className="react-datepicker-wrapper"
        >
          <div
            className="react-datepicker__input-container"
          >
            <input
              autoComplete={undefined}
              autoFocus={undefined}
              className="form-control date-filter"
              disabled={false}
              id={undefined}
              name={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              placeholder="To.."
              readOnly={undefined}
              required={undefined}
              tabIndex={undefined}
              title={undefined}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div />
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Date
          </th>
          <th>
            Distance
          </th>
          <th>
            Time
          </th>
          <th>
            Speed
          </th>
          <th>
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            className="date"
          >
            Sunday, September 3rd 2017
          </td>
          <td>
            100
             m
          </td>
          <td>
            100
             s
          </td>
          <td>
            1
             m/s
          </td>
          <td
            className="actions"
          >
            <a
              className="btn btn-danger"
              onClick={[Function]}
            >
              <i
                className="fa fa-trash"
              />
            </a>
            <a
              className="btn btn-warning"
              onClick={[Function]}
            >
              <i
                className="fa fa-edit"
              />
            </a>
          </td>
        </tr>
        <tr>
          <td
            className="date"
          >
            Wednesday, September 6th 2017
          </td>
          <td>
            200
             m
          </td>
          <td>
            200
             s
          </td>
          <td>
            1
             m/s
          </td>
          <td
            className="actions"
          >
            <a
              className="btn btn-danger"
              onClick={[Function]}
            >
              <i
                className="fa fa-trash"
              />
            </a>
            <a
              className="btn btn-warning"
              onClick={[Function]}
            >
              <i
                className="fa fa-edit"
              />
            </a>
          </td>
        </tr>
      </tbody>
    </table>
    <a
      href="/report"
      onClick={[Function]}
    >
      Report by week
    </a>
    <div
      className="action-footers"
    >
      <a
        className="btn btn-primary"
        onClick={[Function]}
      >
        <i
          className="fa fa-plus"
        />
      </a>
    </div>
  </div>
</div>
`;
